keps/sig-api-machinery/2896-openapi-v3/README.md

```diff
@@ -0,0 +1,563 @@
+
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Kubernetes resources and types can be described through their OpenAPI definitions, currently served as OpenAPI v2 at the $cluster/openapi/v2 cluster endpoint. With the introduction of CRDs, users provide an OpenAPI v3 spec to describe the resource. Since kubernetes only publishes OpenAPI v2, these definitions are converted into an OpenAPI v2 definition before being aggregated and served at the /openapi/v2 endpoint. OpenAPI v3 is more expressive than v2, and the conversion results in the loss of some information. Some of the type definitions are just transformed into “accept anything” definitions because of our lack of ability to perform a better conversion and limitations in kubectl. Some efforts at improving the OpenAPI (eg: enum support) are targeted to bo bundled with OpenAPI v3.
+
+Additionally, since kubernetes aggregate the OpenAPI v2 spec for all specs into one endpoint, whenever a spec is modified the entire spec is updated and all clients will need to redownload the spec. Depending on the size and number of CRDs, this is not an inexpensive operation.
+
+### Goals
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Support publishing and aggregating OpenAPI v3 for all kubernetes types
+- Published v3 spec should be a lossless representation of the data and no fields should be stripped for both built-in types and CRDs
+- Instead of serving the entire OpenAPI spec at one endpoint, separate the spec by group and only serve the resources required by each group
+
+#### Transparency in the OpenAPI
+
+There are a couple of important fields part of the CRD structural schema that are stripped when we publish v2: `oneOf`, `anyOf`, `nullable`, `default`. We would like to keep all fields and publish v3 specs without any information loss for both built-in types and CRDs.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- Accommodating new OpenAPI v3 fields outside the Schema object that do not affect validity of the spec. These are nice to have fields that can be added later.
+- Consumers of the OpenAPI (eg: Server Side Apply, client-go, kubectl, etc) will eventually need to be updated to consume v3. This is outside the scope of this KEP.
```

Comment:
We have two different pruning code-paths/variants afaik. SSA does not use the same slimmed-down spec as kubectl.

Reply:
does client-go consume openapi today? Am not aware of that.

Reply:
I'm not actually sure about the use cases, but the discovery client does have an option to download the openapi.

Reply:
> We have two different pruning code-paths/variants afaik. SSA does not use the same slimmed-down spec as kubectl.

Unfortunately SSA does use a somewhat slimmed-down spec not because of the generation, but because we convert to a v2 proto model. (https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go#L1402) (https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/util/openapi/proto.go#L35)